spec/filters/aggregate_spec.rb

```diff
@@ -0,0 +1,203 @@
+# encoding: utf-8
+require "logstash/devutils/rspec/spec_helper"
+require "logstash/filters/aggregate"
+
+describe LogStash::Filters::Aggregate do
+	def event(data = {})
+		data["message"] ||= "Log message"
+		data["@timestamp"] ||= Time.now
+		LogStash::Event.new(data)
+	end
+
+	def start_event(data = {})
+		data["logger"] = "TASK_START"
+		event(data)
+	end
+
+	def update_event(data = {})
+		data["logger"] = "DAO"
+		event(data)
+	end
+
+	def end_event(data = {})
+		data["logger"] = "TASK_END"
+		event(data)
+	end
+
+	def setup_filter(config = {})
+		config["task_id"] ||= "%{requestid}"
+		filter = LogStash::Filters::Aggregate.new(config)
+		filter.register()
+		return filter
+	end
+
+	def aggregate_maps()
+		LogStash::Filters::Aggregate.aggregate_maps
+	end
+
+	def filter(event)
+		@start_filter.filter(event)
+		@update_filter.filter(event)
+		@end_filter.filter(event)
+	end
+
+	before(:each) do
+		LogStash::Filters::Aggregate.set_eviction_instance_nil()
+		aggregate_maps.clear()
+		@start_filter = setup_filter({ "map_action" => "create", "code" => "map['dao.duration'] = 0" })
+		@update_filter = setup_filter({ "map_action" => "update", "code" => "map['dao.duration'] += event['duration']" })
+		@end_filter = setup_filter({ "map_action" => "update", "code" => "event.to_hash.merge!(map)", "end_of_task" => true, "timeout" => 5 })
+	end
+
+	context "Start event" do
+		describe "and receiving an event without task_id" do
+			it "does not record it" do
+				@start_filter.filter(event())
+				insist { aggregate_maps.size } == 0
+			end
+		end
+		describe "and receiving an event with task_id" do
+			it "records it" do
+				event = start_event("requestid" => "id123")
+				@start_filter.filter(event)
+
+				insist { aggregate_maps.size } == 1
+				insist { aggregate_maps["id123"].nil? } == false
+				insist { aggregate_maps["id123"].creation_timestamp } >= event["@timestamp"]
+				insist { aggregate_maps["id123"].map["dao.duration"] } == 0
+			end
+		end
+
+		describe "and receiving two 'start events' for the same task_id" do
+			it "keeps the first one and does nothing with the second one" do
+
+				first_start_event = start_event("requestid" => "id124")
+				@start_filter.filter(first_start_event)
+				
+				first_update_event = update_event("requestid" => "id124", "duration" => 2)
+				@update_filter.filter(first_update_event)
+				
+				sleep(1)
+				second_start_event = start_event("requestid" => "id124")
+				@start_filter.filter(second_start_event)
+
+				insist { aggregate_maps.size } == 1
```

Comment:
I will try to fix it this week-end. If I need help, I would not hesitate to ask help to you.